### PR TITLE
Add how-to on how to control compositor behavior regarding client decoration requests

### PR DESCRIPTION
**Pull request by tarek-y-ismail**
Adds a small how-to describing how to make the compositor prefer server side decorations as well as mentioning the other alternatives.

### CODE REVIEW

**Review by AlanGriffiths — CHANGES_REQUESTED** [review]
I think this is better presented as a "tutorial" starting from a basic example (something approximating the basic "Anatomy" example in https://mir-server.io/docs/developing-a-wayland-compositor-using-mir) and adding a custom `Decorations` option.

File: doc/sphinx/reference/introducing_the_miral_api.md.in
[the remark this answers is not in the document]
Comment (tarek-y-ismail) [reply doc/sphinx/reference/introducing_the_miral_api.md.in:1]: I agree. I only took this as a starting point. I see two things to do:
1. Re-write the "Introducing the Miral API" page to be a bit more thorough and gradual. Starting with the bare minimum and adding components as we go, mentioning what each component does in more detail (vs  the comments in the code)
2. Adding one of the `miral::Decorations::*` at some point along with a proper explanation as well.

What do you think?

File: doc/sphinx/reference/introducing_the_miral_api.md.in [hunk 1]
```diff
@@ -24,3 +24,18 @@ The way these events are handled determines the behaviour of the shell.
...
+1. {func}`miral::Decorations::always_ssd`: Always uses server side decorations regardless of what the client requests.
+2. {func}`miral::Decorations::prefer_ssd`: Respects the client's requested decoration mode, but uses server side decorations if the client doesn't specify.
+3. {func}`miral::Decorations::always_csd`: Similar to {func}`miral::Decorations::always_ssd`, but always uses client side decorations.
+4. {func}`miral::Decorations::prefer_csd`: Similar to {func}`miral::Decorations::prefer_ssd`, but prefers client side decorations.
+
+You can customize this behavior by implementing a class similar to `ConfigureDecorations` that reads an environment variable, a config file, or any other way you'd like to configure your compositor.
+
+If no strategy is provided in {func}`miral::MirRunner::run_with`, then the default strategy is to prefer client side decorations.
+
+You can also implement your own strategies by implementing the {class}`DecorationStrategy` interface if the need arises.
```
[the remark this answers is not in the document]
Comment (tarek-y-ismail) [reply doc/sphinx/reference/introducing_the_miral_api.md.in:41]: You are correct. I felt like this wasn't really a common use case and your comment now cements this. Will remove!

File: doc/sphinx/reference/introducing_the_miral_api.md.in
[the remark this answers is not in the document]
Comment (AlanGriffiths) [reply doc/sphinx/reference/introducing_the_miral_api.md.in:1]: I think we need a base "Using Mir libraries to build your own Wayland compositor" leading to something approximating the basic "Anatomy" example in https://mir-server.io/docs/developing-a-wayland-compositor-using-mir.

Other tutorials (such as "changing the decoration strategy") can use that as a base.

There is no future in "Introducing the Miral API"

File: doc/sphinx/how-to/how-to-control-decorations.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
+# How to Control Decorations
+A big part of writing your own compositor is to be able to customize its looks,
+behaviour, and effects to your own liking. One way of customizing and unifying
+the looks of applications is server side decorations. MirAL allows you to
+customize how the server deals with decoration requests using
+{class}`miral::Decorations`.
```
Comment (AlanGriffiths) [inline doc/sphinx/how-to/how-to-control-decorations.md:6]: This text will confuse. Users may reasonably expect to customise the decorations (colour, position of buttons, style, ...) 

We need to make it clear that this HOWTO is about choosing CSD or SSD, it isn't about customising decorations.

**Review by AlanGriffiths — CHANGES_REQUESTED** [review]
Need to clarify what this delivers

**Comment by tarek-y-ismail** [comment]
> Need to clarify what this delivers

You mean the PR description?

**Comment by AlanGriffiths** [comment]
> > Need to clarify what this delivers
> 
> You mean the PR description?

title, description & slug

File: doc/sphinx/how-to/how-to-control-compositor-responses-to-client-decoration-requests.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
+# How to Control Compositor Responses to Client Decoration Requests
+This how-to will show you how to specify the behavior of the compositor
+regarding client window decoration requests.
+
+In general, clients can ask the compositor to choose for them, or request
+server or client side decorations from the compositor excplicilty. MirAL allows
+you to customize how the server deals with these requests using
+{class}`miral::Decorations`.
```
Comment (AlanGriffiths) [inline doc/sphinx/how-to/how-to-control-compositor-responses-to-client-decoration-requests.md:8]:
```suggestion
Clients can ask the compositor to use server or client side decorations, or request the compositor to choose for them. {class}`miral::Decorations` allows you to customize how the server deals with these requests.

This how-to will show you how to specify the behavior of the compositor regarding these requests.
```

File: doc/sphinx/write-your-first-wayland-compositor.md [hunk 4]
```diff
@@ -113,3 +113,4 @@ Now that you have your base compositor working, feel free to check out these gui
 
 - [How to specify start up applications](how-to/how-to-specify-startup-apps.md)
 - [How to handle user input](how-to/how-to-handle-keyboard-input.md)
+- [How to Control Compositor Responses to Client Decoration Requests](/doc/sphinx/how-to/how-to-control-compositor-responses-to-client-decoration-requests.md)
```
Comment (AlanGriffiths) [inline doc/sphinx/write-your-first-wayland-compositor.md:116]:
```suggestion
- [How to Control Compositor Responses to Client Decoration Requests](how-to/how-to-control-compositor-responses-to-client-decoration-requests.md)
```

**Review by AlanGriffiths — CHANGES_REQUESTED** [review]
Nits

File: doc/sphinx/how-to/index.md [hunk 5]
```diff
@@ -1,12 +1,15 @@
...
 
 - [Developing a Compositor](developing-a-wayland-compositor-using-mir): build your own, custom compositor with Mir
 - [Developing protocol extensions for Mir](developing-wayland-extension-protocols-for-mir-servers): extend your compositor with support for a custom protocol extension
 - [Getting involved in Mir](getting_involved_in_mir): first steps to contributing to Mir
 - [Using checkbox-mir](how-to-use-checkbox-mir): validating your environment using [checkbox-mir](https://snapcraft.io/checkbox-mir)
 - [Calibrating a touchscreen device](how-to-calibrate-a-touchscreen-device): creating and using a calibration matrix using [the libinput snap](https://snapcraft.io/libinput)
 - [Enabling graphics-core22 on a device](how-to-enable-graphics-core22-on-a-device): creating a `graphics-core22` provider snap for a new board
+- [How to specify start up applications](..//how-to/how-to-specify-startup-apps.md)
+- [How to handle user input](../how-to/how-to-handle-keyboard-input.md)
+- [How to Control Compositor Responses to Client Decoration Requests](../how-to/how-to-control-compositor-responses-to-client-decoration-requests.md)
```
Comment (AlanGriffiths) [inline doc/sphinx/how-to/index.md:12]: This title (also used elsewhere and for slug) is hard to parse.

Perhaps: "Negotiating client/server side decorations" (not great I admit)

**Review by AlanGriffiths — APPROVED** [review]
Take or leave the suggested rename

File: doc/sphinx/how-to/index.md (hunk 5 above)
Comment (Saviq) [reply doc/sphinx/how-to/index.md:12]: I also prefer (as you can see above) "Specifying…", "Handling…" in the link titles, otherwise we'll have "How to" most of the page content :)

You also don't need `../how-to/`.

"Deciding between client / server side decorations"…?
"Preferring client or server side decorations"…?
"Selecting client / server decoration preference"…?

Comment (AlanGriffiths) [reply doc/sphinx/how-to/index.md:12]: Or... can we assume some domain knowledge in anyone looking for this?

"Setting CSD/SSD preference"

Comment (tarek-y-ismail) [reply doc/sphinx/how-to/index.md]: I took a bit of both suggestions :) 

A small note about the new file name. I couldn't bring myself to prefix it with `how-to-` since it read too awkwardly. I don't think it's a big issue other than breaking the uniformity of naming in the how-to directory.

**Review by AlanGriffiths — APPROVED** [review]
Works
